logfds.1
.\" (c) 2014 Richard Kettlewell
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.TH logfds 1
.SH NAME
logfds \- log output of a command
.SH SYNOPSIS
.B logfds
.RB [ -qcC ]
.RB [ -m
.IR days ]
.B --
.IR fd [, fd... ]
.I pattern ...
.B --
.IR command ...
.SH DESCRIPTION
\fBlogfds\fR executes a command in a child process and redirects
output from the specified file descriptors to logfiles.
.PP
The command line should consist of any options desired, followed by a
list of redirections, followed by the command to execute.  Each of
these three sections should be separated by a "--" argument.
.PP
Each redirection consists of two arguments.  The first is a list of
comma-separated file descriptors to redirect, the second is a filename
pattern.  This has the same syntax as the format string of
\fBstrftime\fR(3).
.PP
It is an error to try to redirect the same file descriptor more than
once.
.PP
The program performs the following steps.  Firstly pipes are set up
for each file descriptor that is to be redirected.  The program
forks.  In the child, each file descriptor to be redirected is
replaced with the write end of the appropriate pipe; then the command
is executed.
.PP
In the parent, the write ends of the pipes are closed, and input is
read from the read ends and copied to the appropriate files.  For each
fragment of data, the pattern is reinterpreted and (if the resulting
name has changed) the output file re-opened.  Periodically old files
are compressed or deleted as controlled by the \fB-c\fR and \fB-m\fR
options.
.PP
If the logging process cannot write output to a file immediately
(perhaps because the disc is full, for example) it buffers the data
until it can.
.PP
When all the pipes reach end of file, the parent waits for the child
to terminate.  When it does, it reports its exit status to standard
error, unless either the exit status was 0 or the \fB-q\fR option was
used; then it exits with the exit status of the child or, if the child
was terminated by a signal, 128 plus the signal number.
.SH OPTIONS
.TP
\fB-q\fR, \fB--quiet\fR
Quiet mode.  Suppresses some informational messages.
.TP
\fB-c\fR, \fB--compress\fR
Compress saved logfiles.
.TP
\fB-m\fR \fIdays\fR, \fB--max-log-age\fR \fIdays\fR
Specify the maximum number of days to keep old log files.  By default,
log files are kept forever.
.TP
\fB-C\fR, \fB--log-in-child\fR
Reverses the usual behaviour and does the logging in the child
process; the parent process executes the command.  This is useful
where it is important that the process ID of the command is the same
as the command starts with.
.TP
\fB-h\fR, \fB--help\fR
Show summary of options.
.TP
\fB-V\fR, \fB--version\fR
Show version of program.
.SH AUTHOR
Richard Kettlewell <rjk@greenend.org.uk>


ChangeLog
2005-02-07 17:17:21 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * adverbio.c: -O FD:PATH where the open(PATH, ...) returned a value
          matching FD would unhelpfuly close FD.  Problem reported by Colin
          Watson.
2005-02-07 17:14:08 +0000  Richard Kettlewell  <rjk@greenend.org.uk>

        * with-lock.c: bogus string literal
2003-07-19 17:35:12  Richard Kettlewell  <rjk@greenend.org.uk>

        * check-tested-everything, test-*: check that every executable has
          been tested
        
        * configure.ac: quieten autoconf
        
        * debug.c, utils.h, run-repeatedly.c: new debugging interface
        
        * test-run-repeatedly: fix test
        
2003-07-19 11:26:29  Richard Kettlewell  <rjk@greenend.org.uk>

        * prepare: don't configure if asked not to
        
        * test-run-repeatedly: tests
        
2003-03-08  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.1: added some examples
	
2002-11-28  Richard Kettlewell  <rjk@greenend.org.uk>

	* iobuffer.c: use a single big circular buffer rather than lots of
	small buffers.  This makes more predictable and efficient use of
	memory than the previous strategy.

	Also change --max-buffer/-m to --buffer/-b since it is now the
	exact buffer size.

2002-11-23  Richard Kettlewell  <rjk@greenend.org.uk>

	* More test infrastructure.  Various tests improved.

	* run-repeatedly.c: underscore in filenames is allowed (to match
	run-parts)

2002-11-22  Richard Kettlewell  <rjk@greenend.org.uk>

	* various changes to build on FreeBSD.  daemon (at least) doesn't
	seem to work properly there though.

2002-11-21  Richard Kettlewell  <rjk@greenend.org.uk>

	* alarm.c: report failed setpgid correctly

	* lots of <syscall>_e functions to keep error handling out of the
	main code.  Convert a couple of malloc/realloc calls to the x-
	versions too (for the same reason).

	* more tests

2002-11-20  Richard Kettlewell  <rjk@greenend.org.uk>

	* test-daemon: more tests, including for some of the new features

	* logfds.c: adjust option letters so daemon & logfds are
	consistent.

	* daemon.c: saner error output handling - essentially build part
	of logfds into daemon.

2002-11-18  Richard Kettlewell  <rjk@greenend.org.uk>

	* anagrams.c: use dynamic memory allocation rather than fixed-size
	buffers.  Now there should be no size limits other than available
	memory.

	Also, now you can specify multiple word lists, and duplicate words
	don't cause duplicate anagrams.

2002-11-17  Richard Kettlewell  <rjk@greenend.org.uk>

	* new programs: iobuffer, anagrams

2002-11-11  Richard Kettlewell  <rjk@greenend.org.uk>

	* inplace.c: fix message reported when a subprocess gets a signal

	* alarm: new --verbose option

	* new programs: logfds, bind/accept/connect-socket, pidfile,
	run-as, daemon, run-repeatedly, with-lock

	* switch to new autoconf &c

2001-09-09  Richard Kettlewell  <rjk@greenend.org.uk>

	* configure.in: changed version to 0.6.0

	* added "daemon" program.

2001-08-12  Richard Kettlewell  <rjk@greenend.org.uk>

	* merged "inplace" and "adverbio" into a single package



logfds.c
/* 
   logfds - log output of a command

   Copyright (C) 2001 Richard Kettlewell

   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

*/

#include <config.h>

#include <getopt.h>
#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <errno.h>
#include <stdlib.h>
#include <ctype.h>
#include <sys/wait.h>

#include "utils.h"
#include "logdaemon.h"

/* Option flags and variables */
static struct option const long_options[] =
{
  { "help", no_argument, 0, 'h' },
  { "version", no_argument, 0, 'V' },
  { "compress", no_argument, 0, 'c' },
  { "quiet", no_argument, 0, 'q' },
  { "max-log-age", required_argument, 0, 'm' },
  { "log-in-child", no_argument, 0, 'C' },
  { "day", required_argument, 0, 'D' },
  { 0, 0, 0, 0}
};

/* write a usage message to FP and exit with the specified status */

static void __attribute__((noreturn)) usage(FILE *fp, int exit_status) {
  if(fputs(
"Usage:\n"
"  logfds [options] [--] fd[,fd...] path ... [--] command ...\n"
"\n"
"Options:\n"
"  -c, --compress                        Compress logs\n"
"  -m DAYS, --max-log-age DAYS           Delete old logs\n"
"  -q                                    Quiet mode\n"
"  -C                                    Log in the child, not the parent\n"
"  -h, --help                            Usage message\n"
"  -V, --version                         Version number\n"
, fp) < 0)
    fatale("output error");
  exit(exit_status);
}

int main(int argc, char **argv) {
  int n;
  int max = 0;
  int compress = 0;
  struct fdmap *fds = 0;
  int quiet = 0;
  int loginchild = 0;
  pid_t pid;
  pid_t r;

  setprogname(argv[0]);
  
  while((n = getopt_long(argc, argv, 
			 "hVqcm:D:C",
			 long_options, (int *)0)) >= 0) {
    switch(n) {
    case 'V':
      printf("logfds %s\n", VERSION);
      return 0;

    case 'h':
      usage(stdout, 0);

    case 'm':
      max = atoi(optarg);
      break;

    case 'c':
      compress = 1;
      break;

    case 'q':
      ++quiet;
      break;

    case 'D':
      ld_day = atol(optarg);
      break;

    case 'C':
      loginchild = 1;
      break;
      
    default:
      usage(stderr, 1);
    }
  }
  
  /* process all redirections */
  while(optind < argc
	&& isdigit(argv[optind][0])) {
    char **v;
    int p[2];
    struct logfile *l;
    int first = 1;

    /* check that there's a pattern */
    if(optind + 1 >= argc)
      fatal("missing filename pattern in redirection");
    /* create the pipe */
    pipe_e(p);
    /* make sure the input end is closed in the child */
    cloexec(p[0]);
    v = split(argv[optind], ',');
    if(!*v)
      fatal("empty file descriptor list in redirection");
    for(n = 0; v[n]; ++n) {
      int fd;
      
      if(!v[n][0] || v[n][strspn(v[n], "0123456789")])
	fatal("invalid file descriptor in redirection");
      /* we dup() file descriptors to keep the fdmap_* functions
       * happy */
      fd = first ? p[1] : dup_e(p[1]);
      fdmap_add(&fds, fd, atoi(v[n]));
      free(v[n]);
      first = 0;
    }
    free(v);
    ++optind;
    l = ld_new_logfile(argv[optind]);
    l->rotate = max;
    l->compress = compress;
    ld_new_input(p[0], l);
    ++optind;
  }

  /* skip optional separator */
  if(optind < argc && !strcmp(argv[optind], "--"))
    ++optind;

  /* check that there is a command */
  if(optind >= argc)
    fatal("no command to execute");

  /* launch the subprocess */
  pid = fork_e();
  if(loginchild ? pid != 0 : pid == 0) {
    exiter = _exit;
    /* redirect file descriptors */
    fdmap_map(fds);
    if(execvp(argv[optind], argv + optind) < 0)
      fatale("error executing %s", argv[optind]);
    fatal("execvp succeeded but returned");
  }
  
  /* close writer ends of pipes */
  fdmap_close(fds);
  fdmap_free(fds);

  /* enter the logger event loop */
  if(ld_loop())
    exit(-1);
  
  /* no more inputs; wait for the child to terminate */
  if(!loginchild) {
    r = waitpid_e(pid, &n, 0);
    if(!r)
      fatal("waitpid unexpectedly returned 0");

    /* report exit status if the child reported an exit status */
    if(n && !quiet)
      fprintf(stderr, "%s[%lu] %s\n",
	      argv[optind], (unsigned long)pid, wstat(n));
  }

  /* pass an appropriate exit status to the caller */
  exit(WIFSIGNALED(n) ? 128 + WTERMSIG(n) : WEXITSTATUS(n));
}
